File: cpp/vector/56. 合并区间.cpp
class Solution {
public:
    const int undefined = -1000000000;
    static bool cmp(vector<int> a,vector<int> b) {
        return a[0]<b[0];
    }
    vector<int> getIntervalVector(int start,int stop) {
        vector<int> element;
        element.push_back(start);
        element.push_back(stop);
        return element;
    }
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        vector<vector<int>> result;
        if (intervals.size()==0){
            return result;
        }
        sort(intervals.begin(),intervals.end(),cmp);
        int curStart = undefined;
        int curStop = undefined;
        for(auto interval : intervals){
            if (curStart==undefined && curStop==undefined){
                curStart = interval[0];
                curStop = interval[1];
                continue;
            }
            if (curStop>=interval[0]) {
                curStop = max(curStop,interval[1]);
            }else{
                result.push_back(getIntervalVector(curStart,curStop));
                curStart = interval[0];
                curStop = interval[1];
            }
        }
       if (curStart!=undefined && curStop!=undefined){
            result.push_back(getIntervalVector(curStart,curStop));
        }
        return result;
    }
};

/*
    知识点
    1. vector 自定义排序。
*/

File: 新刷题/数据结构/数组/一维数组/56. 合并区间.go
package 一维数组

import (
	"fmt"
	"sort"
)

// merge 区间合并。
func merge(intervals [][]int) [][]int {
	// 1. 空返回。
	if len(intervals) == 0 {
		return nil
	}

	// 2. 排序。
	sort.Slice(intervals, func(i, j int) bool {
		return intervals[i][0] < intervals[j][0]
	})

	// 3. 贪心。
	result := make([][]int, 0)
	left, right := intervals[0][0], intervals[0][1]
	for i := 1; i < len(intervals); i++ {
		// 3.1 如果当前区间的右端点包括了下一区间的左端点，此时扩展当前区间。
		if right >= intervals[i][0] {
			right = max(right, intervals[i][1])
			continue
		}

		// 3.2 如果当前区间的右端点不包括下一区间的左端点，区间截断。
		result = append(result, []int{left, right})
		left, right = intervals[i][0], intervals[i][1]
	}
	result = append(result, []int{left, right})

	// 4. 返回。
	return result
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
